test/unit/org/apache/cassandra/utils/AbstractTypeGenerators.java

```diff
@@ -0,0 +1,362 @@
+    }
+
+    public static Gen<MapType<?, ?>> getMapType(Gen<AbstractType<?>> typeGen)
+    {
+        return getMapType(typeGen, typeGen);
+    }
+
+    public static Gen<MapType<?, ?>> getMapType(Gen<AbstractType<?>> keyGen, Gen<AbstractType<?>> valueGen)
+    {
+        return rnd -> MapType.getInstance(keyGen.generate(rnd), valueGen.generate(rnd), BOOLEAN_GEN.generate(rnd));
+    }
+
+    public static Gen<TupleType> getTupleType()
+    {
+        return getTupleType(getType(2)); // lower the default depth since this is already a nested type
+    }
+
+    public static Gen<TupleType> getTupleType(Gen<AbstractType<?>> elementGen)
+    {
+        return getTupleType(elementGen, VERY_SMALL_POSITIVE_SIZE_GEN);
+    }
+
+    public static Gen<TupleType> getTupleType(Gen<AbstractType<?>> elementGen, Gen<Integer> sizeGen)
+    {
+        return rnd -> {
+            int numElements = sizeGen.generate(rnd);
+            List<AbstractType<?>> elements = new ArrayList<>(numElements);
+            for (int i = 0; i < numElements; i++)
+                elements.add(elementGen.generate(rnd));
+            return new TupleType(elements);
+        };
+    }
+
+    public static Gen<UserType> getUserType()
+    {
+        return getUserType(getType(2)); // lower the default depth since this is already a nested type
+    }
+
+    public static Gen<UserType> getUserType(Gen<AbstractType<?>> elementGen)
+    {
+        return getUserType(elementGen, VERY_SMALL_POSITIVE_SIZE_GEN);
+    }
+
+    public static Gen<UserType> getUserType(Gen<AbstractType<?>> elementGen, Gen<Integer> sizeGen)
+    {
+        return rnd -> {
+            boolean multiCell = BOOLEAN_GEN.generate(rnd);
+            int numElements = sizeGen.generate(rnd);
+            List<AbstractType<?>> fieldTypes = new ArrayList<>(numElements);
+            LinkedHashSet<FieldIdentifier> fieldNames = new LinkedHashSet<>(numElements);
+            String ks = IDENTIFIER_GEN.generate(rnd);
+            ByteBuffer name = AsciiType.instance.decompose(IDENTIFIER_GEN.generate(rnd));
+
+            // UDTs don't allow duplicate names, so make sure all names are unique
+            for (int i = 0; i < numElements; i++)
+            {
+                fieldTypes.add(elementGen.generate(rnd));
+                // keep creating namings until they are unique
+                while (!fieldNames.add(FieldIdentifier.forQuoted(IDENTIFIER_GEN.generate(rnd)))) {}
+            }
+            return new UserType(ks, name, new ArrayList<>(fieldNames), fieldTypes, multiCell);
+        };
+    }
+
+    public static Gen<AbstractType<?>> allowReversed(Gen<AbstractType<?>> gen)
+    {
+        return rnd -> BOOLEAN_GEN.generate(rnd) ? ReversedType.getInstance(gen.generate(rnd)) : gen.generate(rnd);
+    }
+
+    /**
+     * For a type, create generators for data that matches that type
+     */
+    public static TypeSupport getTypeSupport(AbstractType<?> type)
```

Comment:
nit: We could go...

```suggestion
    public static <T> TypeSupport<T> getTypeSupport(AbstractType<T> type)
```

...but it only pushes the warning down. At least it wouldn't be in the top-level test cases.

Reply:
many annotations later.... done